src/wasm/wasm-type.cpp

```diff
@@ -393,6 +393,14 @@ bool Type::isException() const {
 
 bool Type::isNullable() const {
   if (isBasic()) {
+    if (id == i31ref) {
+      // i31ref is always non-nullable - there is no way to declare a nullable
+      // i31 reference, unlike other reference types. For now, allow it in vars
+      // for testing (as we have no other way to test it, yet), by reporting
+      // that it is nullable.
+      // FIXME when we have proper non-nullability.
+      return true;
+    }
```

Comment:
I thought primitive types are defaultable too, because they can be set to 0, but this does not allow it. [This](https://github.com/WebAssembly/binaryen/blob/2257f857069faa56335d2e24d7d6853c9501fcb7/src/wasm/wasm-validator.cpp#L2282) seems to imply all basic types are defaultable.

Reply:
[The GC MVP doc says i31ref is not](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#reference-types)], at least unless I'm misreading it
```
anyref == (ref null any)

eqref == (ref null eq)

dataref == (ref data)

i31ref == (ref i31)
```
So apparently dataref and i31ref are non-nullable?

Reply:
I wonder if `i31ref` is meant to default to 0? I assume `(ref null i31)` defaults to `null` regardless.

Reply:
Oh sorry nevermind, I thought this method was `isDefaultable`, but it was `isNullable`....

Reply:
Not sure if it helps, but I got [this response](https://github.com/WebAssembly/gc/issues/124#issuecomment-680941832) when asking about `i31ref == (ref i31)` specifically.

> Having a different nullability default for the i31ref shorthand was a pragmatic choice based on expected use cases, but it is a bit irregular. We can switch it around if folks think that's preferable. Opinions?

Potentially related, `i31.new` has recently been promoted to a [constant instruction](https://github.com/WebAssembly/gc/pull/150/files#diff-1df1ea4af362cad5073bac12a65046e4c21bdaa644f69e203b41687b933ac91dR303), which may or may not indicate that `i31ref` can now default to `(i31.new (i31.const 0))`.

Reply:
Thanks @dcodeIO , interesting.

Sounds like not everything is figured out in the spec about this, so we'll revisit as necessary...